Bug 1690342: Initial DNS status conditions implementation

danehans opened the pull request:
Previously, status conditions for the `dns.operator.openshift.io/v1` resource were not implemented.

This commit does the following:

- Implements `Degraded`, `Progressing` and `Available` status conditions for `dns.operator.openshift.io/v1`.
- Refactors cluster-dns-operator `Degraded` status condition to be calculated from the the availability of `ClusterIP`.
- Refactors cluster-dns-operator `Progressing` and `Available` status conditions to be calculated from the `dns.operator.openshift.io/v1` `Available` condition.
- Replaces deployment logic from cluster-dns-operator unit tests with `dns.operator.openshift.io/v1` `Available` conditions.
- Adds `dns.operator.openshift.io/v1` status condition unit tests.

Miciah commented on pkg/operator/controller/status_test.go:
```diff
@@ -7,15 +7,14 @@ import (
-		haveNamespace                           bool
-		numWanted, numAvailable, numUnavailable int
+		haveNamespace           bool
+		numWanted, numAvailable int
```
It feels like we have lost some test coverage here, because whereas before we were counting things that had a positive number in some particular field, now we are counting things that have a slice of things where one of those things has a specific value in one particular field and another specific value in another particular field.  That is to say, there are many more ways the logic could go wrong.  What do you think about adding inputs for the number of dnses without an `Available` condition, number with `Available` set to `Unknown`, and number with `Available` set to `False`, as well as a test case for each of these and for a combination of several?

danehans replied:
Thank you for the review. I have addressed all your comments in my latest commit, except this one. My understanding is `numUnavailable` was removed because the `Progressing` and `Available` status conditions are now based on the `DNS` status condition instead of comparing the number of available and unavailable `DaemonSets`.

Miciah commented on pkg/operator/controller/dns_status_test.go:
```diff
@@ -0,0 +1,258 @@
+	}{
+		{"daemonset pod available on 0/2 nodes", 0, 2, operatorv1.OperatorStatusTypeAvailable, operatorv1.ConditionFalse},
+		{"daemonset pod available on 1/2 nodes", 1, 2, operatorv1.OperatorStatusTypeAvailable, operatorv1.ConditionTrue},
+		{"daemonset pod available on 2/2 nodes", 2, 2, operatorv1.OperatorStatusTypeAvailable, operatorv1.ConditionTrue},
```
We should decide what 0/0 means and add a test case.

Miciah commented on pkg/operator/controller/controller.go:
```diff
@@ -316,34 +316,14 @@ func (r *reconciler) ensureClusterDNS(dns *operatorv1.DNS) error {
-		if err := r.syncClusterDNSStatus(dns, clusterIP, clusterDomain); err != nil {
-			errs = append(errs, fmt.Errorf("failed to sync status of cluster dns %s/%s: %v", daemonset.Namespace, daemonset.Name, err))
+		if err := r.syncDNSStatus(daemonset, dns); err != nil {
+			errs = append(errs, fmt.Errorf("failed to sync dns status: %v", err))
```
`syncDNSStatus` should set `.status.clusterIP` and `.status.clusterDomain`.

danehans commented:
/test images

danehans commented:
Updated PR to also fix [bug 1695210](https://bugzilla.redhat.com/show_bug.cgi?id=1695210). __Note:__ This bug references cluster-ingress-operator, but cluster-dns-operator is affected by the same bug.

danehans commented:
/test e2e-aws-operator

danehans commented:
/test e2e-aws-operator

danehans commented:
Holding until https://github.com/openshift/cluster-ingress-operator/pull/201 merges since dns operator requires the same bug fix.

danehans commented:
Holding until https://github.com/openshift/cluster-ingress-operator/pull/201 merges since dns operator requires the same bug fix.

danehans commented:
/hold

knobunc commented:
@danehans openshift/cluster-ingress-operator#201 has merged... should the hold be lifted?

danehans commented:
@knobunc I would like to see https://github.com/openshift/cluster-dns-operator/pull/92 and https://github.com/openshift/cluster-ingress-operator/pull/200 merge. Both are close.

danehans commented:
/hold cancel

danehans commented:
```
Apr 25 03:56:25.007 E clusteroperator/machine-config changed Degraded to True: error pool master is not ready, retrying. Status: (pool degraded: false total: 3, ready 2, updated: 2, unavailable: 1): Unable to apply 0.0.1-2019-04-25-025503: error pool master is not ready, retrying. Status: (pool degraded: false total: 3, ready 2, updated: 2, unavailable: 1)
Apr 25 04:01:19.105 E clusterversion/version changed Failing to True: ClusterOperatorDegraded: Cluster operator kube-apiserver is reporting a failure
```
/test e2e-aws-upgrade

danehans commented:
@Miciah @pravisankar I pulled the hold and this PR is ready for your review.

ironcladlou commented:
I think this is good. @Miciah PTAL and tag when ready

pravisankar commented on pkg/operator/controller/dns_status.go:
```diff
@@ -0,0 +1,146 @@
+	updated := dns.DeepCopy()
+	updated.Status.ClusterIP = clusterIP
+	updated.Status.ClusterDomain = clusterDomain
+	updated.Status.Conditions = computeDNSStatusConditions(updated.Status.Conditions, clusterIP, ds)
```
We do want to pass the old status conditions, `computeDNSStatusConditions(dns.Status.Conditions,..`  might be clear?

danehans replied:
`setDNSStatusCondition` creates a slice of conditions. The func has logic to append conditions to the new slice and return it. I thought we should be operating on a copy (i.e. `updated`) of the object under these circumstances?

pravisankar commented on pkg/operator/controller/dns_status.go:
```diff
@@ -0,0 +1,146 @@
+	conditions = computeDNSDegradedCondition(conditions, clusterIP)
+	conditions = computeDNSProgressingCondition(conditions, ds)
+	conditions = computeDNSAvailableCondition(conditions, ds)
+
```
`computeDNSStatusConditions` does not depend on old status conditions so we don't have to pass this as an argument?
May be we could do:
```
func computeDNSStatusConditions(ds *appsv1.DaemonSet) []operatorv1.OperatorCondition {
        conditions := []operatorv1.OperatorCondition{}
	conditions = append(conditions, computeDNSDegradedCondition(clusterIP))
	conditions = append(conditions, computeDNSProgressingCondition(ds))
	conditions = append(conditions, computeDNSAvailableCondition(ds))
        return conditions
}
```
Each of the `computeDNSConditions` methods could return a condition and we no longer need the `setDNSStatusCondition` method.

Miciah replied:
`computeDNSFooCondition` needs the old condition as input so that it can return the old condition with the old last transition time if the `Foo` condition did not change.

Miciah replied:
On further thought, we could do something very close to Ravi's suggestion:
```go
func computeDNSStatusConditions(conditions []configv1.ClusterOperatorStatusCondition, ds *appsv1.DaemonSet) []operatorv1.OperatorCondition {
	conditions = setOperatorStatusCondition(conditions, computeDNSDegradedCondition(clusterIP))
	conditions = setOperatorStatusCondition(conditions, computeDNSProgressingCondition(ds))
	conditions = setOperatorStatusCondition(conditions, computeDNSAvailableCondition(ds))
        return conditions
}
```
I kind of like that, and it should correctly preserve or as appropriate update `LastTransitionTime`.

Miciah commented on pkg/operator/controller/dns_status.go:
```diff
@@ -0,0 +1,146 @@
+}
+
+// computeDNSStatusConditions computes dns status conditions
+// based on the status of ds.
```
Technically, that should be "based on the status of ds and clusterIP"... but maybe I am being excessively fussy.

Miciah commented on pkg/operator/controller/dns_status.go:
```diff
@@ -0,0 +1,146 @@
+	}
+	if !cmp.Equal(a.Conditions, b.Conditions, conditionCmpOpts...) || a.ClusterIP != b.ClusterIP || a.ClusterDomain != b.ClusterDomain {
+		return false
+	}
```
In `statusesEqual`, we do the following:
```go
	if foo is different {
		return false
	}
	if bar is different {
		return false
	}
	// ...
	return true
```
I find that approach easier to read.  I don't object if you prefer a single `if foo is different || if bar different || ... {` statement though.  However, if you do go in that direction, you could simplify further by doing away with the `if` and simply returning a Boolean expression:
```go
	return foo is the same && bar is the same && ...
```
Concretely:
```go
	return cmp.Equal(a.Conditions, b.Conditions, conditionCmpOpts...) && a.ClusterIP == b.ClusterIP && a.ClusterDomain == b.ClusterDomain
```

danehans replied:
I agree with your recommendation and find the code easier to read that way. I'll update.

Miciah commented on pkg/operator/controller/dns_status_test.go:
```diff
@@ -0,0 +1,379 @@
+		description                                string
+		clusterIP                                  string
+		avail, desire                              int32
+		degradeType, progressType, availType       string
```
Why are `degradeType`, `progressType`, and `availType` inputs? They are all the same for all test cases, and each one has only one reasonable value.

Miciah commented on pkg/operator/controller/dns_status_test.go:
```diff
@@ -0,0 +1,379 @@
+			availType:      operatorv1.OperatorStatusTypeAvailable,
+			degradeStatus:  operatorv1.ConditionTrue,
+			progressStatus: operatorv1.ConditionTrue,
+			availStatus:    operatorv1.ConditionTrue,
```
This looks odd—is DNS really available if it has no IP address?

danehans replied:
Good point. The previous "available" condition logic only considered having 1 running pod in the daemonset for the condition to be true. Let me update that logic to consider the clusterIP too.

danehans replied:
I will update the available condition to consider the ClusterIP status.

Miciah commented on pkg/operator/controller/dns_status_test.go:
```diff
@@ -0,0 +1,379 @@
+			availType:      operatorv1.OperatorStatusTypeAvailable,
+			degradeStatus:  operatorv1.ConditionFalse,
+			progressStatus: operatorv1.ConditionTrue,
+			availStatus:    operatorv1.ConditionTrue,
```
Would it make sense to say that the service is degraded if it is available on some but not all nodes? It is still available, but for pods on a node where the daemonset is not running, DNS requests will be routed via the service proxy to another node, so performance and possibly reliability may be slightly degraded.

danehans replied:
Yes, I will update the condition to consider the ClusterIP status.

Miciah commented on pkg/operator/controller/status.go:
```diff
@@ -128,58 +129,77 @@ func computeStatusConditions(conditions []configv1.ClusterOperatorStatusConditio
 		degradedCondition.Message = "DNS namespace does not exist"
 	} else {
 		degradedCondition.Status = configv1.ConditionFalse
+		degradedCondition.Reason = "NamespaceExists"
```
It looks like most operators use "AsExpected" for the `Degraded` condition's `Reason` field when its `Status` field is false:
```console
% oc get clusteroperators -o json | jq -r '.items|.[].status.conditions|.[]|select(.type=="Degraded")|[.reason][0]' | sort | uniq -c | sort -n
      1 NoCredentialsFailing
      5 null
     10 AsExpected
```
I'm not one to advocate blind conformity in all things, but I submit the above for your consideration.

danehans replied:
I'll use `AsExpected`.

Miciah commented on pkg/operator/controller/status.go:
```diff
@@ -128,58 +129,77 @@ func computeStatusConditions(conditions []configv1.ClusterOperatorStatusConditio
+	}
+	if numDNSes == dnsesAvailable {
 		progressingCondition.Status = configv1.ConditionFalse
+		progressingCondition.Reason = "AllDaemonSetPodsRunning"
```
Since I provided data on how other operators set the `Degraded` condition, I'll provide similar data on the `Progressing` condition, for your consideration:
```console
% oc get clusteroperators -o json | jq -r '.items|.[].status.conditions|.[]|select(.type=="Progressing")|[.reason][0]' | sort |
uniq -c | sort -n
      1 Ready
      1 ReconcilingComplete
     10 AsExpected
     13 null
```

Miciah commented on pkg/operator/controller/status.go:
```diff
@@ -128,58 +129,77 @@ func computeStatusConditions(conditions []configv1.ClusterOperatorStatusConditio
-	if len(unavailable) == 0 {
+	if dnsesAvailable > 0 {
 		availableCondition.Status = configv1.ConditionTrue
+		availableCondition.Reason = "DNSAvailable"
```
```console
% oc get clusteroperators -o json | jq -r '.items|.[].status.conditions|.[]|select(.type=="Available")|[.reason][0]' | sort | uniq -c | sort -n
      1 MinimumAvailability
     10 AsExpected
     14 null
```

danehans replied:
I'll use `AsExpected`.

Miciah commented on pkg/operator/controller/dns_status.go:
```diff
@@ -0,0 +1,173 @@
+	case len(clusterIP) == 0:
+		degradedCondition.Status = operatorv1.ConditionTrue
+		degradedCondition.Reason = "NoClusterIP"
+		degradedCondition.Message = "No Cluster IP for DNS Service available"
```
Should be "ClusterIP" for consistency (or alternatively we could change the "ClusterIP" to "Cluster IP" above and in `computeDNSAvailableCondition`).  Also, maybe "available"→"assigned".

Miciah commented on pkg/operator/controller/dns_status.go:
```diff
@@ -0,0 +1,173 @@
+	case ds.Status.NumberAvailable != ds.Status.DesiredNumberScheduled:
+		degradedCondition.Status = operatorv1.ConditionTrue
+		degradedCondition.Reason = "DaemonSetDegraded"
+		degradedCondition.Message = "All nodes not running DaemonSet pod"
```
This could be misinterpreted as saying that no nodes are running the DaemonSet pod.  "Not all nodes running" or "Some nodes not running" would be clearer.

Miciah commented on pkg/operator/controller/status.go:
```diff
@@ -95,29 +90,35 @@ func (r *reconciler) syncOperatorStatus() error {
+	dnses []operatorv1.DNS) []configv1.ClusterOperatorStatusCondition {
+	conditions = computeOperatorDegradedCondition(conditions, ns)
+	conditions = computeOperatorProgressingCondition(conditions, dnses)
+	conditions = computeOperatorAvailableCondition(conditions, dnses)
```
Could we use the same `conditions = setStatusCondition(conditions, computeFooCondition(...))` pattern here that you implemented in `computeDNSStatusConditions`?

danehans replied:
Sure. I'll push an update to the PR in a few minutes. Thanks for all the great feedback!

Miciah commented:
/lgtm
/hold
@pravisankar, do you want to give this another look before we merge? If you're fine with the changes, feel free to cancel the hold.

pravisankar commented on pkg/operator/controller/dns_status_test.go:
```diff
@@ -0,0 +1,354 @@
+			desire:         0,
+			degradeStatus:  operatorv1.ConditionTrue,
+			progressStatus: operatorv1.ConditionFalse,
+			availStatus:    operatorv1.ConditionFalse,
```
As @ironcladlou suggested on the ingress pr https://github.com/openshift/cluster-ingress-operator/pull/200 , this can be simplified:
```
type conditions struct {
	degraded, progressing, available bool
}
And then in the test case, we could do
...
expectedConditions: conditions{true, false, false}
...
```

pravisankar commented on pkg/operator/controller/status_test.go:
```diff
@@ -25,48 +24,92 @@ func TestComputeStatusConditions(t *testing.T) {
+				degraded:    true,
+				progressing: true,
+				available:   false,
+			},
```
This can be converted to compact format.

pravisankar commented on pkg/operator/controller/dns_status.go:
```diff
@@ -0,0 +1,173 @@
+func computeDNSStatusConditions(conditions []operatorv1.OperatorCondition, clusterIP string, ds *appsv1.DaemonSet) []operatorv1.OperatorCondition {
+	conditions = setDNSStatusCondition(conditions, computeDNSDegradedCondition(clusterIP, ds))
+	conditions = setDNSStatusCondition(conditions, computeDNSProgressingCondition(ds))
+	conditions = setDNSStatusCondition(conditions, computeDNSAvailableCondition(clusterIP, ds))
```
We could get rid of `setDNSStatusCondition`. Instead of looping all conditions for every `computeDNSXYZCondition` to check for changes, we could compute actual condition once and can patch the LastTransitionTime. Also did not like the fact that LastTransitionTime for the condition is set outside of `computeDNSXYZCondition` method.
WDYT of  https://github.com/openshift/cluster-ingress-operator/pull/200/files#diff-31d813e7a1281abc26f3e03bc7686530R163 ?

pravisankar commented on pkg/operator/controller/status.go:
```diff
@@ -95,29 +90,35 @@ func (r *reconciler) syncOperatorStatus() error {
+	dnses []operatorv1.DNS) []configv1.ClusterOperatorStatusCondition {
+	conditions = setOperatorStatusCondition(conditions, computeOperatorDegradedCondition(ns))
+	conditions = setOperatorStatusCondition(conditions, computeOperatorProgressingCondition(dnses))
+	conditions = setOperatorStatusCondition(conditions, computeOperatorAvailableCondition(dnses))
```
We could get rid of setOperatorStatusCondition and can set LastTransitionTime as part of computeOperatorXYZCondition.

pravisankar commented on pkg/operator/controller/dns_status_test.go:
```diff
@@ -0,0 +1,354 @@
+		description   string
+		oldConditions []operatorv1.OperatorCondition
+		newCondition  *operatorv1.OperatorCondition
+		expected      []operatorv1.OperatorCondition
```
Instead of using operatorv1.OperatorCondition, we could use conditions{available: true}, conditions{degraded: false, progressing: true} and build the actual conditions based on this concise input before calling the test method.

danehans commented:
@pravisankar thank you for the review. I have updated the PR based on your comments.

Miciah commented on pkg/operator/controller/status.go:
```diff
@@ -95,123 +99,184 @@ func (r *reconciler) syncOperatorStatus() error {
+	}
 
-	newConditions := []configv1.ClusterOperatorStatusCondition{}
+	return true
```
We decided that a DaemonSet is available if 0/0 replicas are available (https://github.com/openshift/cluster-dns-operator/pull/88#discussion_r272673377, https://github.com/openshift/cluster-dns-operator/commit/5ac65e290667f001e1029fb0f69dc359875451c6#diff-c7469ee2a512d7eff942a36312cb4c50R20), and I see we are extending that reasoning by saying that all DNSes are available if 0/0 DNSes are available.

However, consider that the operator will automatically recreate the `default` DNS if it does not exist.  This consideration leads me to think that having 0/0 DNSes implies that `Progressing` is true, and possibly that `Available` is false.  We have discussed the analogous case of 0/0 ingress controllers in relation to the ingress operator: https://github.com/openshift/cluster-ingress-operator/pull/198#issuecomment-481058325.

ironcladlou replied:
Given our invariant that `default` must always exist, agree that 0 `dns` resources means progressing=true.

IMO available should always be false if the actual replicas of `default` is 0 regardless of what the "desired". I don't understand how, from our operator's perspective, DNS could be considered available if no CoreDNS pods are running.

danehans replied:
>Given our invariant that default must always exist, agree that 0 dns resources means progressing=true.

Done, latest PR includes `Progressin=true` if `if dnsesAvailable == 0 || dnsesAvailable != dnsesTotal`.

pravisankar replied:
When there are no dns resources, I was expected these conditions:
Availabe = False --> No DNS is running and serving the requests
Progressing = False --> Operator is not actually doing any work or trying to transition from one state to another.
Degraded = True --> Operator expects default dns to be present but does not exist which is a failure.
@ironcladlou @Miciah What do you think?

Miciah replied:
> Availabe = False --> No DNS is running and serving the requests

I agree.

> Progressing = False --> Operator is not actually doing any work or trying to transition from one state to another.

I disagree because the operator calls `ensureDefaultDNS` in a periodic goroutine, so the operator is doing work (or will be doing work in less than 1 minute) to create the default resource.

> Degraded = True --> Operator expects default dns to be present but does not exist which is a failure.

I agree.

pravisankar replied:
That makes sense. Okay, progressing=true when there are no dns resources.

Miciah commented on pkg/operator/controller/status.go:
```diff
@@ -95,123 +99,184 @@ func (r *reconciler) syncOperatorStatus() error {
-				return oldConditions
+// checkAnyDNSesAvailable checks if all DNS resources are available.
+func checkAnyDNSesAvailable(dnses []operatorv1.DNS) bool {
+	var dnsesAvailable int
```
Note for follow-up: This can be a Boolean value.  (Alternatively, if we decide not to combine `checkAnyDNSesAvailable` and `checkAllDNSesAvailable` as I suggested in another comment, we can do away with the variable and just `dnsesAvailable++`→`return true` below.)

Miciah commented on pkg/operator/controller/status.go:
```diff
@@ -95,123 +99,184 @@ func (r *reconciler) syncOperatorStatus() error {
-		availableCondition.Reason = "DaemonSetNotAvailable"
-		availableCondition.Message = strings.Join(unavailable, "\n")
+		availableCondition.Reason = "DNSUnavailable"
+		availableCondition.Message = fmt.Sprintf("No dns resources available")
```
Note for follow-up: Use consistent capitalization of "dns".

danehans commented on pkg/operator/controller/status.go:
```diff
@@ -44,13 +45,16 @@ func (r *reconciler) syncOperatorStatus() error {
 	}
 
+	allDNSesAvailable := checkAllDNSesAvailable(dnses)
+	anyDNSesAvailable := checkAnyDNSesAvailable(dnses)
```
I converted both funcs into a `numDNSesAvailable` func and compare the number of dnses available to the total number of dnses in the status conditions calculation.

danehans commented on pkg/operator/controller/status.go:
```diff
@@ -95,123 +99,184 @@ func (r *reconciler) syncOperatorStatus() error {
+}
+
+// computeOperatorStatusConditions computes the operator's current state.
+func (r *reconciler) computeOperatorStatusConditions(oldConditions []configv1.ClusterOperatorStatusCondition, ns *corev1.Namespace,
```
I have removed the receiver. It was added to include status versions as part of the `Progressing` condition calculation. The version calculation can be included in https://github.com/openshift/cluster-dns-operator/pull/92

Miciah commented on pkg/operator/controller/status.go:
```diff
@@ -95,123 +94,147 @@ func (r *reconciler) syncOperatorStatus() error {
+	conditions := []configv1.ClusterOperatorStatusCondition{}
+	conditions = append(conditions, computeOperatorDegradedCondition(oldDegradedCondition, dnsesAvailable, dnses, ns))
+	conditions = append(conditions, computeOperatorProgressingCondition(oldProgressingCondition, dnsesAvailable, dnses))
+	conditions = append(conditions, computeOperatorAvailableCondition(oldAvailableCondition, dnsesAvailable))
```
In my earlier comment about using a slice literal, I meant to suggest the following:
```go
	conditions := []configv1.ClusterOperatorStatusCondition{
		computeOperatorDegradedCondition(oldDegradedCondition, dnsesAvailable, dnses, ns),
		computeOperatorProgressingCondition(oldProgressingCondition, dnsesAvailable, dnses),
		computeOperatorAvailableCondition(oldAvailableCondition, dnsesAvailable),
	}
￼```

Miciah commented on pkg/operator/controller/status.go:
```diff
@@ -95,123 +94,147 @@ func (r *reconciler) syncOperatorStatus() error {
 	}
-	if ns == nil {
+	switch {
+	case dnsesAvailable == 0 || dnsesAvailable != dnsesTotal && ns == nil:
```
`&&` is higher precedence than `||`, so this expression groups as follows: `dnsesAvailable == 0 || (dnsesAvailable != dnsesTotal && ns == nil)`.

Miciah replied:
Since the expression starts to get hairy with the three subexpressions, I'd suggest simplifying by putting `case ns == nil:` first so that subsequent cases can assume the `ns` is not nil.  If the namespace does not exist, then DNSes cannot exist, so I believe it is sufficient just to say that the namespace does not exist if that is the case.

pravisankar replied:
Good catch

Miciah commented on pkg/operator/controller/status.go:
```diff
@@ -15,24 +13,23 @@ import (
-	DNSClusterOperatorName = "dns"
+	DNSOperatorName     = "dns"
+	OperatorVersionName = "operator"
+	DNSVersionName      = DNSOperatorName
```
This is no longer used in this PR, but no harm in leaving it in since #92 will most likely use it.

Miciah commented:
/lgtm

Miciah commented:
level=error msg="\t* module.vpc.aws_route.to_nat_gw[4]: 1 error occurred:"
    level=error msg="\t* aws_route.to_nat_gw.4: Error creating route: timeout while waiting for state to become 'success' (timeout: 2m0s)"

/test e2e-aws-upgrade

ironcladlou commented:
/retest

ironcladlou commented:
/hold cancel

knobunc commented:
/hold cancel
